docs/samples/python/example_pdf_annotation.ipynb

```diff
@@ -0,0 +1,254 @@
+    "            analyzer_results = analyzer.analyze(text=text_to_anonymize, language='en')\n",
+    " \n",
+    "            if text_to_anonymize.isspace() == False:\n",
+    "                print(text_to_anonymize)\n",
+    "                print(analyzer_results)\n",
+    "\n",
+    "            characters = list([])\n",
+    "\n",
+    "            # Grab the characters from the PDF\n",
+    "            for text_line in filter(lambda t: isinstance(t, LTTextLine), text_container):\n",
+    "                    for character in filter(lambda t: isinstance(t, LTChar), text_line):\n",
+    "                            characters.append(character)\n",
+    "\n",
+    "\n",
+    "            # Slice out the characters that match the analyzer results.\n",
+    "            for result in analyzer_results:\n",
+    "                start = result.start\n",
+    "                end = result.end\n",
+    "                analyzed_character_sets.append({\"characters\": characters[start:end], \"result\": result})"
+   ]
+  },
+  {
+   "attachments": {},
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Create phrase bounding boxes\n",
+    "\n",
+    "The next task is to take the character data, and inflate it into full phrase bounding boxes.\n",
+    "\n",
+    "For example, for an email address, we'll turn the bounding boxes for each character in the email address into one single bounding box."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# Combine the bounding boxes into a single bounding box.\n",
+    "def combine_rect(rectA, rectB):\n",
+    "    a, b = rectA, rectB\n",
+    "    startX = min( a[0], b[0] )\n",
+    "    startY = min( a[1], b[1] )\n",
+    "    endX = max( a[2], b[2] )\n",
+    "    endY = max( a[3], b[3] )\n",
+    "    return (startX, startY, endX, endY)\n",
+    "\n",
+    "analyzed_bounding_boxes = []\n",
+    "\n",
+    "# For each character set, combine the bounding boxes into a single bounding box.\n",
+    "for analyzed_character_set in analyzed_character_sets:\n",
+    "    completeBoundingBox = analyzed_character_set[\"characters\"][0].bbox\n",
+    "    \n",
+    "    for character in analyzed_character_set[\"characters\"]:\n",
+    "        completeBoundingBox = combine_rect(completeBoundingBox, character.bbox)\n",
+    "    \n",
+    "    analyzed_bounding_boxes.append({\"boundingBox\": completeBoundingBox, \"result\": analyzed_character_set[\"result\"]})"
+   ]
+  },
+  {
+   "attachments": {},
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Add highlight annotations\n",
+    "\n",
+    "We finally iterate through all the analyzed bounding boxes and create highlight annotations for all of them."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "pdf = Pdf.open(\"./sample_data/sample.pdf\")\n",
```

Comment:
Is it possible to read the pdf once and use the same object throughout the sample notebook?

Reply:
So we choose to use two PDF libraries here, the first is pdfminer, which we use solely for text extraction, we do so as it's sole function is to extract data (text) from a PDF accurately. In the case of this sample, the text extraction needs to be accurate so we've got the best chance of finding PII. For example, some libraries 'break' the text when it's extracted, leading to extra or missing spaces in text. Unfortunately text extraction from a PDF is not a trivial thing to get right, but pdfminer seems to do the best job of it.

The second library we use is pikepdf. We choose this one for updating and writing out as pdfminder doesn't have writing capabilities, and of the other PDF libraries, this one is meant to produce good, valid PDF. It also supports standards such as PDF/A, which is used quite a lot in government and organisations where archiving is important.

So long story short, we open the file twice, once for each library. :)

Reply:
I see, thanks!